grpclog/internal/loggerv2.go

```diff
@@ -190,15 +211,35 @@ type LoggerV2Config struct {
 // The infoW, warningW, and errorW writers are used to write log messages of
 // different severity levels.
 func NewLoggerV2(infoW, warningW, errorW io.Writer, c LoggerV2Config) LoggerV2 {
-	var m []*log.Logger
 	flag := log.LstdFlags
 	if c.FormatJSON {
 		flag = 0
 	}
-	m = append(m, log.New(infoW, "", flag))
-	m = append(m, log.New(io.MultiWriter(infoW, warningW), "", flag))
-	ew := io.MultiWriter(infoW, warningW, errorW) // ew will be used for error and fatal.
-	m = append(m, log.New(ew, "", flag))
-	m = append(m, log.New(ew, "", flag))
+
+	infoLogW := infoW
+
+	// Use io.Discard instead of io.MultiWriter when all loggers at a given level
+	// are set to io.Discard. Both this package and the standard log package have
+	// significant optimizations for io.Discard, which io.MultiWriter lacks (as of
+	// this writing).
+
+	warnLogW := io.Discard
+	if warningW != io.Discard || infoW != io.Discard {
+		warnLogW = io.MultiWriter(infoW, warningW)
+	}
+
+	errLogW := io.Discard
+	if errorW != io.Discard || warningW != io.Discard || infoW != io.Discard {
+		errLogW = io.MultiWriter(infoW, warningW, errorW)
+	}
```

Comment:
I think we could optimize this a bit, and make it easier to read as well?

```go
// Returns a combined logger for both higher & lower severity logs, or higher only if lower is io.Discard.
func combineLoggers(lower, higher io.Writer) io.Writer {
	if lower == io.Discard || higher == io.Discard {
		return higher
	}
	return io.MultiWriter(lower, higher)		
}

func NewLoggerV2(..) {
	// ...

	// Just overwrite instead of creating new locals with different names:
	warningW = combineLoggers(infoW, warningW)
	errorW = combineLoggers(warningW, errorW)
}
```

This ends up with 2 nested multiwriters if you have 3 different loggers, but that's probably fairly uncommon, and would be no worse than the behavior this PR adds of creating a multiwriter for `(warningW, io.Discard)` if `infoW` is `io.Discard`.

Reply:
This would be a behavioural change. It's currently possible to set a lower level without a higher one, e.g. log info but not warning. With that change, any existing user doing that would find their info being discarded. Is that acceptable?

Reply:
Ok, I chanaged it to:
```
if lower == io.Discard { 
	return higher 
} 
if lower == io.Discard { 
	return higher 
}
```

Which I believe preserves existing behaviour, if that still accomplishes what you want?

> // Just overwrite instead of creating new locals with different names:

I removed the locals, except `fatalW := errorW` to make the code self-documenting. But I can remove that too if you want.

> This ends up with 2 nested multiwriters if you have 3 different loggers

MultiWriter is smart enough to avoid this:
https://github.com/golang/go/blob/946b4baa/src/io/multi.go#L127-L131